File: verification/ht.c
#include <assert.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define FNV_OFFSET 14695981039346656037UL
#define FNV_PRIME 1099511628211UL

uint64_t hash_key_c(const char* key) {
    uint64_t hash = FNV_OFFSET;
    for (const char* p = key; *p; p++) {
        uint64_t val = (uint64_t)(unsigned char)(*p);
        hash ^= val;
        hash *= FNV_PRIME;
    }
    return hash;
}

File: verification/ht.rs
use std::os::raw::c_char;
use std::ffi::CString;

const FNV_OFFSET: u64 = 14695981039346656037;
const FNV_PRIME: u64 = 1099511628211;

#[no_mangle]
fn hash_key_rust(key: *const c_char) -> u64 {
    let mut hash = FNV_OFFSET;
    let mut p = key;

    unsafe {
        while *p != 0 {
            hash ^= (*p as u8) as u64;
            hash = hash.wrapping_mul(FNV_PRIME);
            p = p.add(1);
        }
    }

    hash
}

#[cfg(kani)]
#[kani::proof]
fn main() {
    // Create a Rust string
    let input = "example";

    // Convert the Rust string into a C-compatible string
    let c_string = CString::new(input).expect("CString::new failed");

    // Call the external function
    unsafe {
        let result = hash_key_rust(c_string.as_ptr());
        println!("Hash result for '{}': {}", input, result);
    }
}